Support decimal quantities in Orders, and fix shipping labels handling of them

hichamboushaba opened the pull request:
Closes #3808 and #4118.

Please don't merge before the FluxC's PR is merged: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2037

#### Testing
Please install the plugin [WPC Product Quantity for WooCommerce](https://wordpress.org/plugins/wpc-product-quantity/) in your store to use decimal quantities.

1. Create an order with decimal quantity > 1 in your store.
2. Go to order details in the app.
3. **Confirm** that you can see the correct quantity.
4. Click on "Create shipping label"
5. Pass the steps until you reach the packaging step.
6. **Confirm** that the items in the package are duplicated depending on their quantity.
7. Purchase the shipping label.
8. Open order details in the web client, and click on "View Details" of the label (https://d.pr/i/AmY0CH).
9. **Confirm** that the items match the correct quantity.
10. Go back to the order details in the app.
11. Click on "Issue a refund"
12. **Confirm** that the max quantity you can select is matching the integer part of the quantity.
13. Issue a refund with full max quantity.
14. In the order details: **Confirm** that the remaining quantity is correct, and that "issue a refund" is not visible

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

hichamboushaba committed:
Change shipping package item's quantity to Int

hichamboushaba committed:
Duplicate items in the packaging step to reflect their quantity

hichamboushaba committed:
Make order's decimal

hichamboushaba committed:
Update refund availability logic to check quantities

hichamboushaba committed:
Bump fluxC hash

hichamboushaba committed:
Fix tests

hichamboushaba committed:
Duplicate product ids to reflect their quantity in the SL purchase API

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/model/Order.kt:
```diff
@@ -64,11 +64,12 @@ data class Order(
 
     @IgnoredOnParcel
-    val availableRefundQuantity = items.sumBy { it.quantity }
+    val isRefundAvailable = refundTotal < total && availableRefundQuantity > 0
```
I changed the logic here to avoid showing "Issue a refund" button if we can't handle the refund in the app, this happens if the quantity is `null` or if it's < 1.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/ShippingLabelRepository.kt:
```diff
@@ -178,7 +178,7 @@ class ShippingLabelRepository @Inject constructor(
                 serviceName = rate.serviceName,
                 carrierId = rate.carrierId,
-                products = labelPackage.items.map { it.productId }
+                products = labelPackage.items.map { item -> List(item.quantity) { item.productId } }.flatten()
```
We identified the issue of the previous implementation just today: p1624622537119300/1624620001.117600-slack-C6H8C3G23

anitaa1990 replied:
np: it would be great to maybe add a comment here to better understand the logic behind this change.

hichamboushaba replied:
Done 8702f46

hichamboushaba committed:
Add a release notes entry

anitaa1990 reviewed (commented):
The code changes look good @hichamboushaba! I have a few questions after testing though.

- When I have a product with quantity < 0, then that product is not eligible for refund. Is that correct? I found it a bit confusing that the product was not displayed in the refund list, even though it's displayed in the order detail.
- The quantity displayed in the refund screen is integers and not decimals.

#### Screenshot of order detail + refund screen
<img src="https://user-images.githubusercontent.com/22608780/123909342-7d926500-d996-11eb-9591-f847f68fa399.png" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/123909345-7ec39200-d996-11eb-93ee-edea0c902292.png" width="300"/>

hichamboushaba commented:
Thanks for the review Anitaa.
> When I have a product with quantity < 0, then that product is not eligible for refund. Is that correct? I found it a bit confusing that the product was not displayed in the refund list, even though it's displayed in the order detail.

This PR didn't make any change to this logic, and for me, the product isn't even displayed in the products section of the order details (because of [this](https://github.com/woocommerce/woocommerce-android/blob/5361df283ab23291e71795b6d5d693622b16cd69/WooCommerce/src/main/kotlin/com/woocommerce/android/model/Refund.kt#L105)). IMO we shouldn't spend a lot of time nor think a lot about negative quantities, as they are not supported in Woo Core, and refunding them using quantities in wp-admin doesn't work.

> The quantity displayed in the refund screen is integers and not decimals.

As we only support integer quantities for the refund, otherwise it would require bigger changes to the refund UI, I make sure the quantity passed to refund in integer, I'll see if changing the one displayed there to decimal while keeping the one used for refund is easy.

hichamboushaba committed:
Merge branch 'develop' into issue/4118-sl-packages-quantities

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/model/Order.kt

hichamboushaba committed:
Update refund logic to display correct item quantity

We made maxQuantity a Float, and added a new property that displays the actual refundable quantity (integer)

hichamboushaba committed:
Move release entry to 7.1

hichamboushaba commented:
@anitaa1990 the commit 868985641f2963809e39e8dde06e0bec416d2ef2 updates the logic to allow displaying decimal quantities in refunds, while allowing refunding only integer values, please check it and let me know what you think

<img width="366" alt="Screen Shot 2021-06-30 at 11 14 35" src="https://user-images.githubusercontent.com/1657201/123944200-cc3eff80-d994-11eb-9108-9ba50d952bb1.png">

anitaa1990 reviewed (approved):
This works! Thanks @hichamboushaba! Definitely better than before :shipit:

anitaa1990 commented:
I'll wait for you to update the FluxC PR after merge and then merge this PR 👍

hichamboushaba committed:
Bump FluxC hash

hichamboushaba committed:
Add comment explaining the products duplication for sl purchase API
